Bug 1407037 - declare things from the provisioner

djmitche opened the pull request.

djmitche commented:
I have already added the necessary scopes for staging and production

helfi92 commented on config.yml:
```diff
@@ -95,7 +99,9 @@ production:
-    id: staging-aws
+    provisionerId: staging-aws
+    description: |
+      Staging instance for the AWS provisioner.  Not for production use.
```
Extra space before "Not"

djmitche replied:
`I     hate you   .` ;)

djmitche replied:
hahah, even in monospace markdown strips consecutive spaces.  See, it's markdown, it doesn't matter.

jhford commented on src/main.js:
```diff
@@ -119,6 +119,27 @@ let load = loader({
+        description: cfg.app.description,
+      });
+
+      await WorkerType.scan({}, {
```
This is unneeded.  We know that this call has happened when the worker type was created, so we shouldn't need to call this again.

djmitche replied:
The intent is to "refresh" this data periodically (here, roughly daily based on dyno cycling).  Similar to how we "refresh" the docs, schemas, and references.

jhford commented on src/api-v1.js:
```diff
@@ -319,6 +322,9 @@ api.declare({
     workerType: workerType,
   });
 
+  let queue = new taskcluster.Queue({credentials: this.credentials});
```
In order to match the existing structure of the provisioner, `queue` should be a module in the main.js and passed in as context to the API.

djmitche replied:
I passed it to WorkerType, at which point the API didn't need it anymore.

jhford replied:
Did you mean to delete this line?  We don't use this queue instance any longer from what I can tell, but maybe I'm wrong.

jhford commented on src/worker-type.js:
```diff
@@ -779,6 +780,18 @@ WorkerType.prototype.determineCapacityChange = function(runningCapacity, pending
+WorkerType.prototype.declareWorkerType = function({queue}) {
+  // keep the worker around well over 24 hours, as we will likely re-declare
+  // at least that often
+  const expires = taskcluster.fromNow('36 hours');
```
Why is there an expires field?  We derive that information from the worker type being seen or not in the queue.

djmitche replied:
Because the workerType might not be seen if no workers are provisioned.

jhford commented on src/worker-type.js:
```diff
@@ -779,6 +780,18 @@ WorkerType.prototype.determineCapacityChange = function(runningCapacity, pending
+  return queue.declareWorkerType(this.provisionerId, this.workerType, {
+    stability: 'stable',
+    expires,
+    description: `${this.description}\n\nOwner: ${this.owner}`,
```
not blocking this review, but maybe an owner field would be merited in the queue method?

djmitche replied:
Yeah, I was between that and removing the owner field from the provisioner's WorkerTypes table, and this seemed like a compromise :)

djmitche replied:
(since none of the owner fields contain anything especially meaningful)

djmitche commented on src/worker-type.js:
```diff
@@ -779,6 +780,18 @@ WorkerType.prototype.determineCapacityChange = function(runningCapacity, pending
   return capacityChange;
 };
 
+WorkerType.prototype.declareWorkerType = function({queue}) {
```
Thanks -- I wasn't sure that was a better idea, but I'll do that.

djmitche commented:
I think that accounts for all of the comments..

jhford commented on src/main.js:
```diff
@@ -379,8 +385,20 @@ let load = loader({
+        });
+
+        await WorkerType.scan({}, {
+          handler: wt => wt.declareWorkerType(),
```
If this really is something which is wanted and it's supposed to be done at an interval, I'd rather express that by putting it in a taskcluster-lib-iterate and have it run periodically instead of relying exclusively on the 24-hour reboot.

djmitche replied:
This follows the model of docs, schemas, and references.

Does lib-iterate guarantee an immediate call, or does it wait `waitTime` ms first?

djmitche replied:
It does :) https://github.com/taskcluster/taskcluster-lib-iterate/pull/5

djmitche commented:
Updated

jhford commented on src/main.js:
```diff
@@ -379,8 +385,33 @@ let load = loader({
+          waitTime: day,
+          monitor,
+          handler: async (watchdog, state) => {
+            watchdog.stop();
```
any reason for this?  The watchdog will through in this case if the handler doesn't complete in 24hours.  I suspect if this code takes 24, something is drastically wrong.

djmitche replied:
I think that's OK -- the idea is that it doesn't really matter how much time this takes.

jhford commented on src/main.js:
```diff
@@ -379,8 +385,33 @@ let load = loader({
+    setup: async ({cfg, provisioner, server, queue, WorkerType, monitor}) => {
+      if (cfg.app.publishQueueMetadata) {
+        const day = 24 * 60 * 60 * 1000;
+        let i = new Iterate({
```
Can you move this to a top level component?

djmitche commented:
TC-lib-iterate's docs suggest it takes care of the error handling and reporting to sentry -- `maxFailures` suggests it catches a failure (rather than crash the process0, and `monitor` suggests that it will report that failure to sentry.  I will tune the `waitTimeAfterFail` to something shorter rather than wait 24h, though.

djmitche commented:
(continuing) Have I misunderstood your comment or how lib-iterate works?

djmitche commented:
I moved to a top-level component and removed the `watchdog.stop()`.  I didn't add `waitTimeAfterFail` after all, since tc-lib-iterate doesn't implement it.  I didn't change any of the error handling, for the reasons above.
